source/HardwareSerial.cpp

```diff
@@ -31,276 +61,559 @@ HardwareSerial::~HardwareSerial()
+    //    m_cancellationTokenSource->get_token())
+    //.wait();
+}
+
+int HardwareSerial::peek(void)
+{
+    int returnByte = NO_PEEK_BYTE;
+
+    EnterCriticalSection(&m_readBufferListLock);
+    if (m_peekByte == NO_PEEK_BYTE)
+    {
+        m_peekByte = read();
     }
-    return _storage[_storageIndex];
+    returnByte = m_peekByte;
+    LeaveCriticalSection(&m_readBufferListLock);
+
+    return returnByte;
 }    
 
 int HardwareSerial::read(void)
 {
-    if (_storageUsed)
-    {
-        if (_storageIndex + 1 >= _storageCount)
-        {
-            _storageUsed = false;
-        }
-        return _storage[_storageIndex++];
-    }
-    else
-    {
-        if (!ReadFile(_comHandle, &_storage, 64, &_storageCount, NULL))
-        {
-#ifdef _DEBUG
-            Log("Error %d when reading file: ", GetLastError());
-            LogLastError();
-#endif
-            return -1;
-        }
+    IBuffer^ buffer;
+    int returnByte = NO_PEEK_BYTE;
 
-        _storageUsed = true;
-        _storageIndex = 0;
+    EnterCriticalSection(&m_readBufferListLock);
 
-        // If only one character was read, then reset it
-        if (_storageIndex + 1 >= _storageCount)
+    if (m_peekByte == NO_PEEK_BYTE)
+    {
+        while ((m_reader == nullptr) && (m_readBufferList.size() > 0))
         {
-            _storageUsed = false;
+            buffer = m_readBufferList.front();
+            m_reader = DataReader::FromBuffer(buffer);
+
+            if (m_reader->UnconsumedBufferLength == 0)
+            {
+                m_readBufferList.pop_front();
+                m_reader = nullptr;
+            }
         }
 
-        // if nothing was read, return 0 and dont do anything
-        if (_storageCount == 0)
+        if (m_reader != nullptr)
         {
-            return -1;
-        }
+            returnByte = (int)m_reader->ReadByte();
 
-        return _storage[_storageIndex++];
+            if (m_reader->UnconsumedBufferLength == 0)
+            {
+                m_readBufferList.pop_front();
+                m_reader = nullptr;
+            }
+        }
+    }
+    else
+    {
+        returnByte = m_peekByte;
+        m_peekByte = NO_PEEK_BYTE;
     }
+
+    LeaveCriticalSection(&m_readBufferListLock);
+
+    return returnByte;
 }
     
 size_t HardwareSerial::write(uint8_t c)
 {
-    DWORD bytesWritten;
+    m_dataWriter->WriteByte(c);
 
-    WriteFile(_comHandle, &c, sizeof(uint8_t), &bytesWritten, NULL);
-    return bytesWritten;
+    concurrency::create_task(m_dataWriter->StoreAsync())
+    .wait();
+
+    return 1;
 }
 
 size_t HardwareSerial::write(const uint8_t *buffer, size_t size)
 {
-    DWORD bytesWritten;
+    Platform::Array<uint8_t>^ dataArray = ref new Platform::Array<uint8_t>(size);
+    for (size_t i = 0; i < size; i++) { dataArray[i] = buffer[i]; }
+
+    m_dataWriter->WriteBytes(dataArray);
+
+    concurrency::create_task(m_dataWriter->StoreAsync(), m_cancellationTokenSource->get_token())
+    .wait();
```

Comment:
I do not expect this will be running on the GUI thread.  The begin function should be safe on the GUI thread, but this one should be protected by our Setup/Loop logic.  We may want to consider making this "GUI thread" safe in the future.